cupy/cudnn.pyx

```diff
@@ -2042,13 +2098,77 @@ def batch_normalization_forward_training(
         # (instead of variance) to resultSaveInvVariance argument. The
         # current implementation of our BN depends on this behavior so that
         # we can reduce the number of reduction kernels.
-        cudnn.batchNormalizationForwardTraining(
-            handle, cudnn_mode, one, zero,
-            x_desc, x.data.ptr, x_desc, y.data.ptr,
-            derivedBnDesc, gamma.data.ptr,
-            beta.data.ptr, factor, running_mean_tmp.data.ptr,
-            running_var_tmp.data.ptr, eps,
-            save_mean.data.ptr, save_inv_std.data.ptr)
+
+        if _cudnn_version >= 7401:
+
+            bn_ops = cudnn.CUDNN_BATCHNORM_OPS_BN
+
+            if (
+                    cudnn_mode == cudnn.CUDNN_BATCHNORM_SPATIAL_PERSISTENT
+                    and x.dtype == numpy.float16
+                    and d_layout == cudnn.CUDNN_TENSOR_NHWC
+                    and x.shape[3] % 4 == 0  # C mod 4 == 0
+            ):
+
+                # Faster NHWC kernel can be triggered by allocating extra
+                # spaces.
+                # https://docs.nvidia.com/deeplearning/sdk/cudnn-archived/cudnn_741/cudnn-developer-guide/index.html#cudnnBatchNormalizationForwardTrainingEx  # NOQA
+                workspace_size = (
+                    cudnn.getBatchNormalizationForwardTrainingExWorkspaceSize(
+                        handle,
+                        cudnn_mode,
+                        bn_ops,
+                        x_desc,  # x
+                        x_desc,  # z
+                        x_desc,  # y
+                        derivedBnDesc,
+                        0,  # activation desc
+                    ))
+                workspace = memory.alloc(workspace_size)
+
+                reserve_space_size = (
+                    cudnn.getBatchNormalizationTrainingExReserveSpaceSize(
+                        handle,
+                        cudnn_mode,
+                        bn_ops,
+                        0,  # activation desc
+                        x_desc,
+                    ))
+                reserve_space = memory.alloc(reserve_space_size)
```

Comment:
From the documentation
```
This workspace is not required to be clean. Moreover, the workspace does not have to remain unchanged between the forward and backward pass, as it is not used for passing any information.
```
Maybe we could avoid allocating memory each time if the size remains unchanged?
I think that the overhead should be pretty small, so maybe the current implementation is simpler and better.

Reply:
Yes, CuPy uses memory pool. Allocation is cheap.